openquake/calculators/classical.py

```
@@ -151,43 +155,43 @@ def classical(sources, sitecol, rlzs_assoc, monitor):
     gsims = rlzs_assoc.gsims_by_trt_id[trt_model_id]
-    for smodel in rlzs_assoc.csm_info.source_models:
-        for trt_model in smodel.trt_models:
-            if trt_model.id == trt_model_id:
-                sm_id = smodel.ordinal
-                break
 
+    dic = AccumDict()
+    dic.siteslice = slice(siteidx, siteidx + len(sitecol))
     if monitor.oqparam.poes_disagg:
-        bbs = [BoundingBox(sm_id, sid) for sid in sitecol.sids]
+        sm_id = rlzs_assoc.get_sm_id(trt_model_id)
+        dic.bbs = [BoundingBox(sm_id, sid) for sid in sitecol.sids]
     else:
-        bbs = []
+        dic.bbs = []
     # NB: the source_site_filter below is ESSENTIAL for performance inside
     # hazard_curves_per_trt, since it reduces the full site collection
     # to a filtered one *before* doing the rupture filtering
     curves_by_gsim = hazard_curves_per_trt(
         sources, sitecol, imtls, gsims, truncation_level,
         source_site_filter=source_site_distance_filter(max_dist),
-        maximum_distance=max_dist, bbs=bbs, monitor=monitor)
-    dic = dict(calc_times=monitor.calc_times, bbs=bbs)
+        maximum_distance=max_dist, bbs=dic.bbs, monitor=monitor)
+    dic.calc_times = monitor.calc_times  # added by hazard_curves_per_trt
     for gsim, curves in zip(gsims, curves_by_gsim):
         dic[trt_model_id, str(gsim)] = curves
     return dic
 
 
-def agg_dicts(acc, val):
+def expand(array, n, aslice):
     """
-    Aggregate dictionaries of hazard curves by updating the accumulator
+    Expand a short array to size n by adding zeros outside of the slice
     """
-    for key in val:
-        if key == 'calc_times':
-            acc[key].extend(val[key])
-        elif key == 'bbs':
-            for bb in val['bbs']:
-                acc['bb_dict'][bb.lt_model_id, bb.site_id].update_bb(bb)
-        else:  # aggregate curves
-            acc[key] = agg_curves(acc[key], val[key])
-    return acc
+    if len(array) == n:  # nothing to expand
```

If `len(array) > n`, I guess we should raise an error